### PR TITLE
Dpl378 duplex tnano seq change sample progression

### PR DESCRIPTION
**Pull request by andrewsparkes**
Sequencescape changes for Story https://github.com/sanger/limber/issues/1077

### CODE REVIEW

File: app/controllers/api/v2/bait_libraries_controller.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+module Api
+  module V2
+    # Provides a JSON API controller for BaitLibraryResource
+    # See: http://jsonapi-resources.com/ for JSONAPI::Resource documentation
+    class BaitLibrariesController < JSONAPI::ResourceController
+      # By default JSONAPI::ResourceController provides most the standard
+      # behaviour, and in many cases this file may be left empty.
+    end
+  end
+end
```
Comment (KatyTaylor) [inline app/controllers/api/v2/bait_libraries_controller.rb:12]: Is this needed, adding bait library to the API? Bait library is set as part of the request isn't it (the thing we were working on together)? But maybe you are also separately reading / creating bait libraries, from Limber, independently of a request?

File: app/resources/api/v2/bait_library_resource.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+module Api
+  module V2
+    # Provides a JSON API representation of BaitLibrary
+    # See: http://jsonapi-resources.com/ for JSONAPI::Resource documentation
+    class BaitLibraryResource < BaseResource
+      # Constants...
+
+      # immutable # uncomment to make the resource immutable
```
Comment (KatyTaylor) [inline app/resources/api/v2/bait_library_resource.rb:10]: Does it need to be mutable?

File: app/resources/api/v2/shared_behaviour/receptacle.rb [hunk 3]
```diff
@@ -38,6 +38,8 @@ module Api::V2::SharedBehaviour::Receptacle
     # Attributes
     attribute :uuid, readonly: true
     attribute :name, delegate: :display_name, readonly: true
+
+    # TODO: these can be removed later, moved onto request_metadata
```
Comment (KatyTaylor) [inline app/resources/api/v2/shared_behaviour/receptacle.rb:42]: What needs to be done here? Data migration? Worth creating a story for it?

File: app/controllers/api/v2/bait_libraries_controller.rb (hunk 1 above)
Comment (KatyTaylor) [reply app/controllers/api/v2/bait_libraries_controller.rb:12]: OK I see you're accessing it in the Limber PR,  `app/models/labware_creators/pcr_cycles_binned_plate/csv_file/row.rb`

Could be set to immutable though potentially?

File: app/controllers/api/v2/bait_libraries_controller.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+module Api
+  module V2
```
[the remark this answers is not in the document]
Comment (andrewsparkes) [reply app/controllers/api/v2/bait_libraries_controller.rb:4]: What's this comment on about bot??

File: app/resources/api/v2/bait_library_resource.rb (hunk 2 above)
Comment (andrewsparkes) [reply app/resources/api/v2/bait_library_resource.rb:10]: What does mutable mean in terms of API v2? Nothing can be updated via v2 right?
I am just doing a lookup, so I guess I can set it.

File: app/controllers/api/v2/bait_libraries_controller.rb (hunk 1 above)
Comment (andrewsparkes) [reply app/controllers/api/v2/bait_libraries_controller.rb:12]: Yeah I have to check the bait library name they put in the customer file matches a real bait library.
In terms of the controller it's the same as all the other v2 controllers.

File: app/models/attributable/attribute.rb [hunk 5]
```diff
@@ -22,7 +22,7 @@ def initialize(owner, name, options = {})
     end
 
     def from(record)
-      record[name]
+      record[name] || record.send(name)
```
[the remark this answers is not in the document]
Comment (andrewsparkes) [reply app/models/attributable/attribute.rb:25]: ok, added

File: app/resources/api/v2/shared_behaviour/receptacle.rb (hunk 3 above)
Comment (andrewsparkes) [reply app/resources/api/v2/shared_behaviour/receptacle.rb:42]: There are 4 receptacle columns that could be removed.
The data I guess could be moved into the library requests, but the store and request metadata attributes are specific to the new DilutionAndCleanup request which will not exist for those old runs.
Will make a backlog story.

File: app/resources/api/v2/bait_library_resource.rb (hunk 2 above)
Comment (KatyTaylor) [reply app/resources/api/v2/bait_library_resource.rb]: Yes, some things can be updated via API v2

File: app/resources/api/v2/shared_behaviour/receptacle.rb (hunk 3 above)
Comment (KatyTaylor) [reply app/resources/api/v2/shared_behaviour/receptacle.rb:42]: Hmm I'm not sure we should remove columns if there's nowhere to put the existing data?

**Review by KatyTaylor — APPROVED** [review]

**Comment by andrewsparkes** [comment]
Closed as work deprecated.

**Pull request closed without merging**